### PR TITLE
Fix excluded task issue for GradleBuild

### LINKED ISSUE #18196: Task exclusion on `GradleBuild` tasks does not work past Gradle version 6.7.1
### Summary
Per the [docs](https://docs.gradle.org/current/userguide/command_line_interface.html#sec:excluding_tasks_from_the_command_line), one should be able to exclude a task from execution using the `-x` flag:
```
./gradlew someTask -x someDependentTask
```

The above would execute `someTask` while not executing `someDependentTask` even if `someTask.dependsOn(someDependentTask)`.

This still works for general tasks, but no longer works with tasks being run within a `GradleBuild` task. 

```kotlin
tasks {
    val first by creating {
        doLast {
            println("first")
        }
    }

    val second by creating {
        doLast {
            println("second")
        }

        dependsOn(first)
    }

    task<GradleBuild>("both") {
        // ensure to pass excluded tasks
        startParameter = gradle.startParameter.newInstance()

        tasks = listOf(
            first.name,
            second.name
        )

        doLast {
            println("both")
        }
    }
}
```

### Expected Behavior

On version 6.7.1:
```
./gradlew second -x first
// second

./gradlew both -x second
// first
// both
```

### Current Behavior

Post version 6.7.1:
```
./gradlew second -x first
// second

./gradlew both -x second
// first
// second
// both
```

### Context

This method of excluding tasks is heavily relied on in an open source versioning tool I help maintain for Gradle:
https://github.com/intuit/auto

The Gradle plugin for Auto uses the [ResearchGate Gradle Release Plugin](https://github.com/researchgate/gradle-release) under the hood to perform the actual version upgrade. The release plugin exposes a top-level `GradleBuild` task, `release` configured to execute the release. However, with the way Auto is structured, we need to ensure all commits are done by Auto, so we exclude certain tasks from `release`, such as the `commitNewVersion` task.

https://github.com/intuit/auto/issues/2068

### Steps to Reproduce 

I made a repo to help exemplify this issue:
https://github.com/sugarmanz/gradle-exclude-task-issue

### Your Environment

Build scan URL: https://scans.gradle.com/s/f7fk6s27kt7ua

### PR DESCRIPTION
**Pull request by jbartok**
Fixes #18196

### CODE REVIEW

**Comment by jbartok** [comment]
@bot-gradle test RFM

**Comment by jbartok** [comment]
@bot-gradle test RFM

**Comment by jbartok** [comment]
@bot-gradle test RFM

**Comment by jbartok** [comment]
@bot-gradle test RFM

File: subprojects/core/src/main/java/org/gradle/execution/CompositeAwareTaskSelector.java [hunk 1]
```diff
@@ -51,13 +50,9 @@ public CompositeAwareTaskSelector(GradleInternal gradle, BuildStateRegistry buil
...
             }
         }
-        // Exclusion didn't match an included build, so it might be a subproject of the root build or a relative path
-        if (gradle.isRootBuild()) {
-            return getUnqualifiedBuildSelector().getFilter(path);
-        } else {
-            // Included build ignores this exclusion since it doesn't apply directly to it
-            return Specs.satisfyAll();
-        }
+        return getUnqualifiedBuildSelector().getFilter(path);
```
Comment (big-guy) [inline subprojects/core/src/main/java/org/gradle/execution/CompositeAwareTaskSelector.java:53]: This is going to have consequences with included builds too and common task names like `jar`.

Take a look at this sample:
https://docs.gradle.org/current/samples/sample_sharing_convention_plugins_with_build_logic.html

If you run `gradle help -x jar` with these changes. The build will fail without running the jar task to generate the plugin:
```
* What went wrong:
Could not apply requested plugin [id: 'myproject.java-conventions'] as it does not provide a plugin with id 'myproject.java-conventions'. This is caused by an incorrect plugin implementation. Please contact the plugin author(s).
> Plugin with id 'myproject.java-conventions' not found.
```

Previously, it was not possible to run tasks from an included build, so being able to exclude tasks didn't really mean anything. We also have it so that you cannot use unqualified task selection with included build tasks, so making the fuzzy match work with excludes makes this inconsistent.

Is there some way we could...
1. Allow the StartParameter in the `GradleBuild` task to be configured programmatically such that excludes work. If someone wants to emulate the old behavior, they could pass exclusions to the task.
2. Or make excludes work in a similar way to included builds. So if the GradleBuild task is called "foo", you have to exclude tasks from that build with `-x :foo:someTask`
3. Or make unqualified excludes apply to builds created via `GradleBuild` in the same way as the root build (without affecting other included builds).

File: subprojects/core/src/main/java/org/gradle/execution/CompositeAwareTaskSelector.java [hunk 2]
```diff
@@ -51,13 +50,9 @@ public CompositeAwareTaskSelector(GradleInternal gradle, BuildStateRegistry buil
...
+
+        //todo: turn issue #18196 into an integration test, which is fixed by this change (manually confirmed)
```
[the remark this answers is not in the document]
Comment (jbartok) [reply subprojects/core/src/main/java/org/gradle/execution/CompositeAwareTaskSelector.java:55]: Nope, I forgot about it... Will do.

File: subprojects/core/src/main/java/org/gradle/execution/CompositeAwareTaskSelector.java (hunk 1 above)
Comment (jbartok) [reply subprojects/core/src/main/java/org/gradle/execution/CompositeAwareTaskSelector.java:53]: So... I re-watched some of Adam's Tech Sessions and armed with some more knowledge I tried to dig further. 

I tried to figure out why the `GradleBuild` bits don't show up as an included build when [CompositeAwareTaskSelector.getFilter()](https://github.com/gradle/gradle/blob/ae3758aa7698196c03e7290116ea4b13388986c0/subprojects/core/src/main/java/org/gradle/execution/CompositeAwareTaskSelector.java#L45) is being executed. 

What I've found was that [CompositeAwareTaskSelector.getFilter()](https://github.com/gradle/gradle/blob/ae3758aa7698196c03e7290116ea4b13388986c0/subprojects/core/src/main/java/org/gradle/execution/CompositeAwareTaskSelector.java#L45) runs much sooner (and on a different thread) than [NestedRootBuildRunner.createNestedBuildTree](https://github.com/gradle/gradle/blob/ae3758aa7698196c03e7290116ea4b13388986c0/subprojects/core/src/main/java/org/gradle/internal/build/NestedRootBuildRunner.java#L40). 

Basically, when we try to build the task filters, the `GradleBuild` part has not been evaluated yet, we aren't aware of it yet. Tackling this feels like a much bigger can of worms than the original issue...

@big-guy @adammurdoch , can you guys pls. advise? Should we go further and if so, then how?

**Comment by ykhandelwal913** [comment]
@jbartok @big-guy @adammurdoch just wanted to follow up here. Our teams would like to update to latest gradle version  but cannot due to this behaviour as this is breaking our process. Please let me know if there is any plan to address this.

**Comment by jbartok** [comment]
Problem seems to have been fixed in the meantime, works fine in `7.6` and later.
